torch/_guards.py

```diff
@@ -0,0 +1,136 @@
+import enum
+from typing import List, Optional, Callable
+import dataclasses
+
+class GuardSource(enum.Enum):
+    LOCAL = 0
+    GLOBAL = 1
+    LOCAL_NN_MODULE = 2
+    GLOBAL_NN_MODULE = 3
+    CONSTANT = 4
+    RANDOM_VALUE = 5
+    SHAPE_ENV = 6
+
+    def select(self, locals_, globals_):
+        if self in (GuardSource.LOCAL, GuardSource.LOCAL_NN_MODULE):
+            return locals_
+        if self in (GuardSource.GLOBAL, GuardSource.GLOBAL_NN_MODULE):
+            return globals_
+        raise NotImplementedError()
+
+    def is_nn_module(self) -> bool:
+        return self in (GuardSource.GLOBAL_NN_MODULE, GuardSource.LOCAL_NN_MODULE)
+
+    def is_local(self):
+        return self in (GuardSource.LOCAL, GuardSource.LOCAL_NN_MODULE)
+
+@dataclasses.dataclass
+class Guard:
+    # The name of a Guard specifies what exactly it is the guard is guarding
+    # on.  The meaning of the name is dependent on the create_fn; you must
+    # look at the use-site inside create_fn to know what name means.
+    #
+    # That being said, although you might think this is just a "name", name is
+    # usually an arbitrary Python expression that will be evaluated with all
+    # globals (and locals, if you create a LOCAL guard) to extract the Python
+    # object that we want to perform guard tests on.  This evaluation
+    # typically happens in GuardBuilder.eval.  In these cases, name is
+    # typically produced by Source.name() (not to be confused with
+    # GuardSource)--morally, we could have stored a Source here.
+    #
+    # Occasionally, name is not a valid Python expression; sometimes
+    # it is meaningless.  Example create_fns that are like this include
+    # GRAD_MODE and SYMBOL_MATCH.
+    name: str
+    source: GuardSource
+    create_fn: Callable[["GuardBuilder", "Guard"], None]
```

Comment:
so what's the plan for dealing with the GuardBuilder layering problem. Mypy will not be happy until you figure something out.

Reply:
We should drop this to be any callable, nothing about this create_fn is tied to GuardBuilder directly anymore. For the sake of layering, maybe just do Callable[Any, None]. 

Or we move the GuardBuilder functions down to this layer, but I dont wanna do that, at least not yet.

Reply:
Made it a base class